jcl/src/java.base/share/classes/jdk/crac/Context.java

```diff
@@ -0,0 +1,83 @@
+/*[INCLUDE-IF CRAC_SUPPORT]*/
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2024
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+package jdk.crac;
+
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+
+import openj9.internal.criu.InternalCRIUSupport;
+
+/**
+ * A Resource that allows other Resources to be registered with it.
+ */
+public abstract class Context<R extends Resource> {
+
+	final static boolean debug = System.getProperty("enable.jdk.crac.api.debug") != null; //$NON-NLS-1$
+
+	static Context<Resource> globalContext = new CRIUSupportContext<>();
+
+	/**
+	 * @param resource a Resource to be registered
+	 */
+	public abstract void register(R resource) throws Exception;
+
+	static Context<Resource> createGlobablContext() {
+		return globalContext;
+	}
+
+	public void checkpointJVM() {
+		globalContext.checkpointJVM();
+	}
+}
+
+class CRIUSupportContext<R extends Resource> extends Context<R> {
+	private InternalCRIUSupport internalCRIUSupport = new InternalCRIUSupport(Paths.get("cpData")) //$NON-NLS-1$
+			.setLeaveRunning(false).setShellJob(true).setFileLocks(true);
```

Comment:
we should add a cmdline option `-XX:CRaCCheckpointTo=[dir]`, we can save `dir` in `J9CRIUCheckpointState` and create a native in InternalCRIUSupport to retrieve the dir from java code. This will be what we set for the dir here.

Reply:
should throw a startup failure if both `-XX:+enableCRIUSupport` and `-XX:CRaCCheckpointTo` are specified.

However, `-XX:CRaCCheckpointTo` should internally enable `J9VM_CRIU_IS_CHECKPOINT_ENABLED` and `J9VM_CRIU_IS_CHECKPOINT_ALLOWED`

Reply:
> we should add a cmdline option -XX:CRaCCheckpointTo=[dir], we can save cracDir in J9CRIUCheckpointState and create a native in InternalCRIUSupport to retrieve the dir from java code. This will be what we set for the dir here.

Do we need a way to specify `setLeaveRunning()`, `setShellJob()`, `setFileLocks()`, `setExtUnixSupport()`, `setLogLevel()`, `setLogFile()`, or `setUnprivileged()`?

I see a few options:
1. add comdline options such as `-XX:CRaCCheckpointTo=[dir]`;
2. specify system properties like `-DCRaCCheckpointTo=[dir]`;
3. add setter APIs within `jdk.crac.Context` though this introduces incompatibilities with openjdk CRaC;
4. add static setter APIs within `openj9.internal.criu.InternalCRIUSupport` for CRaC assuming only one `InternalCRIUSupport` instance is required, i.e., the `Context.globalContext`

@tajila any preferences?

Reply:
> Do we need a way to specify setLeaveRunning(), setShellJob(), setFileLocks(), setExtUnixSupport(), setLogLevel(), setLogFile(), or setUnprivileged()?

For now no, we can leave those as default. Once we have more real world use cases we can modify the defaults if needed.

> I see a few options:

it should be `-XX:CRaCCheckpointTo=[dir]` to match existing expectations. We can add more in the future if needed. We should maintain consistency with existing Crac apis.

Reply:
Implemented `-XX:CRaCCheckpointTo=[dir]` and also throw a startup failure if both `-XX:+enableCRIUSupport` and `-XX:CRaCCheckpointTo` are specified.